### ISSUE TITLE
Inconsistency between iterator concepts and functional concepts

### ISSUE THREAD

**Issue by cjdb** [ericniebler/stl2#336]
I find it confusing that the iterator concepts live in `<experimental/ranges/iterator>`, but the functional (or invocable) concepts are not in `<experimental/ranges/functional>`. Adding to this confusion is the fact that invocable proto-concepts are available in `<range/v3/utility/functional.hpp>`, although this probably isn't a major point of contention.

Is it worth discussing whether to merge iterator concepts into `<experimental/ranges/concepts>`, or (preferably) to merge invocable concepts into `<experimental/ranges/functional>`?

**Comment by CaseyCarter** [comment]
The callable concepts characterize the behavior of `invoke`; it makes sense to me that they should live alongside it.

The `<experimental/ranges/concepts>` header feels odd to me in general, much like having a header named `<classes>` or `<function_templates>` would. We don't make headers in C++ by grouping entities by their syntactic type, we group components by functionality. I understand how we ended up here - the concepts are the new "weird" thing - but looking at the subclauses of the "Concepts Library" I'm thinking that "Callable concepts" should be alongside `invoke`, and the others belong in `<type_traits>`.

(I'm inclined to put the Callable concepts and `invoke` in `<type_traits>` as well, but that would disagree with the C++17 WP. That change should probably come from "upstream.")

**Comment by ericniebler** [comment]
> I understand how we ended up here - the concepts are the new "weird" thing - but looking at the subclauses of the "Concepts Library" I'm thinking that "Callable concepts" should be alongside `invoke`, and the others belong in `<type_traits>`.

I see what you're saying, but putting a concept like `Semiregular` in `<type_traits>` also feels wrong to me. It is *not* a type trait -- at least, not as traits have traditionally been known in C++. No suggestions at the moment, just voicing my concerns.

**Comment by CaseyCarter** [comment]
> I see what you're saying, but putting a concept like `Semiregular` in `<type_traits>` also feels wrong to me. 

`Regular` et al aren't type traits, but they definitely fit under the "type property" umbrella of [meta.unary.prop] along with `is_copy_constructible` and `is_nothrow_destructible` et al. I suppose I'm not really suggesting that `<type_traits>` is the proper header so much as that it's sensible for `Semiregular` to live with the type property traits.

**Comment by ericniebler** [comment]
> I suppose I'm not really suggesting that `<type_traits>` is the proper header so much as that it's sensible for `Semiregular` to live with the type property traits.

Is there any practical difference?

**Comment by CaseyCarter** [comment]
> Is there any practical difference?

No, and that's half the point: we have a header named `<type_traits>` whose name is a mechanism that aggregates together several components that use that mechanism. `<experimental/ranges/concepts>` is a header that similarly aggregates components according to mechanism rather than usage. `Semiregular`  doesn't really belong in `<type_traits>` but it does belong with `is_nothrow_destructible`. They should maybe go in a future `<type_properties>` header.

These thoughts are much more speculation than specification, but I could see us reasonably moving the `Invocables` alongside `invoke` in the TS time frame.

**Comment by CaseyCarter** [comment]
PR - in the OP - desperately seeking attention.

**Comment by ericniebler** [comment]
Subclause? IIUC, it goes: clause, section, subsection, subsubsection, etc.

**Comment by timsong-cpp** [comment]
No, it's "part" (which we don't have), then "clause", then "subclause", "subclause", "subclause", "subclause", "subclause", "paragraph".
